File: CLI/src/screenshot.ts
import { connect, ConnectionOptions } from "./discovery.js";
import {
  DEFAULT_HIERARCHY_OPTIONS, Flag, Rect, ViewNode, DeviceInfo,
} from "./protocol.js";

function area(r: Rect): number {
  return Math.max(0, r.width) * Math.max(0, r.height);
}

/** A node worth trying to render: on-screen, visible, and non-empty. */
function renderable(node: ViewNode): boolean {
  return (node.flags & (Flag.hidden | Flag.zeroSize)) === 0
    && node.opacity > 0.001
    && area(node.frame) > 0;
}

/**
 * Ordered list of node ids to try when capturing the whole screen. Prefers the
 * largest visible root window, then its largest first-level children as a
 * fallback — a UIWindow renders directly on iOS, whereas an NSWindow root is not
 * itself a view, so its contentView (a child) is what actually renders on macOS.
 */
export function pickScreenshotNodes(roots: ViewNode[]): string[] {
  const ids: string[] = [];
  const byAreaDesc = (a: ViewNode, b: ViewNode) => area(b.frame) - area(a.frame);
  const sortedRoots = roots.filter(renderable).sort(byAreaDesc);
  for (const root of sortedRoots) {
    ids.push(root.id);
    const kids = root.children.filter(renderable).sort(byAreaDesc).slice(0, 3);
    for (const k of kids) ids.push(k.id);
  }
  return [...new Set(ids)];
}

export interface ScreenCapture {
  png: Buffer;
  nodeID: string;        // the node that actually rendered
  device: DeviceInfo;
}

/**
 * Capture a PNG of the current screen in one round-trip. Fetches the hierarchy,
 * picks the best renderable root (see {@link pickScreenshotNodes}), and renders
 * the first candidate that produces an image. SwiftUI reflection is skipped — we
 * only need view geometry to choose what to render, which keeps it fast.
 */
export async function captureScreen(
  conn: ConnectionOptions,
  scale: number,
): Promise<ScreenCapture> {
  const { client, info } = await connect(conn);
  try {
    const snapshot = await client.fetchHierarchy({
      ...DEFAULT_HIERARCHY_OPTIONS,
      includeSwiftUI: false,
      includeLayers: false,
    });
    const candidates = pickScreenshotNodes(snapshot.roots);
    if (candidates.length === 0) {
      throw new Error("no visible on-screen window to capture (is the app foregrounded?)");
    }
    let lastErr: unknown;
    for (const id of candidates) {
      try {
        const png = await client.fetchSnapshot(id, scale);
        return { png, nodeID: id, device: info.device };
      } catch (e) { lastErr = e; }
    }
    const reason = lastErr instanceof Error ? lastErr.message : String(lastErr);
    throw new Error(`could not capture the screen (tried ${candidates.length} root node(s)): ${reason}`);
  } finally {
    client.disconnect();
  }
}

File: python/src/treescope_cli/screenshot.py
"""Whole-screen capture. Mirrors ``CLI/src/screenshot.ts``."""

from __future__ import annotations

from typing import Any

from .discovery import ConnectionOptions, connect
from .protocol import DEFAULT_HIERARCHY_OPTIONS, Flag


def _area(r: dict) -> float:
    return max(0.0, r.get("width", 0)) * max(0.0, r.get("height", 0))


def _renderable(node: dict) -> bool:
    return (
        (node.get("flags", 0) & (Flag.hidden | Flag.zeroSize)) == 0
        and node.get("opacity", 1) > 0.001
        and _area(node.get("frame", {})) > 0
    )


def pick_screenshot_nodes(roots: list[dict]) -> list[str]:
    """Ordered node ids to try when capturing the whole screen.

    Prefers the largest visible root window, then its largest first-level
    children as a fallback — a UIWindow renders directly on iOS, whereas an
    NSWindow root is not itself a view, so its contentView (a child) renders on
    macOS.
    """
    ids: list[str] = []
    sorted_roots = sorted((r for r in roots if _renderable(r)), key=_area_of_frame, reverse=True)
    for root in sorted_roots:
        ids.append(root["id"])
        kids = sorted((c for c in root.get("children", []) if _renderable(c)), key=_area_of_frame, reverse=True)
        for kid in kids[:3]:
            ids.append(kid["id"])
    # de-duplicate, preserving order
    seen: set[str] = set()
    out: list[str] = []
    for i in ids:
        if i not in seen:
            seen.add(i)
            out.append(i)
    return out


def _area_of_frame(node: dict) -> float:
    return _area(node.get("frame", {}))


async def capture_screen(opts: ConnectionOptions, scale: float) -> dict[str, Any]:
    """Capture a PNG of the current screen in one round-trip.

    Returns ``{"png": bytes, "node_id": str, "device": dict}``.
    """
    client, info = await connect(opts)
    try:
        snapshot = await client.fetch_hierarchy({
            **DEFAULT_HIERARCHY_OPTIONS,
            "includeSwiftUI": False,
            "includeLayers": False,
        })
        candidates = pick_screenshot_nodes(snapshot.get("roots", []))
        if not candidates:
            raise RuntimeError("no visible on-screen window to capture (is the app foregrounded?)")
        last_err: Exception | None = None
        for node_id in candidates:
            try:
                png = await client.fetch_snapshot(node_id, scale)
                return {"png": png, "node_id": node_id, "device": info["device"]}
            except Exception as exc:
                last_err = exc
        raise RuntimeError(
            f"could not capture the screen (tried {len(candidates)} root node(s)): {last_err}"
        )
    finally:
        await client.disconnect()
